tests/ui/author/for_loop.stdout

```diff
@@ -33,31 +40,176 @@ if_chain! {
     if let PatKind::Struct(ref path6, ref fields1, false) = arms1[0].pat.kind;
     if matches!(path6, QPath::LangItem(LangItem::OptionSome, _));
     if fields1.len() == 1;
-    // unimplemented: field checks
+    if fields1[0].ident.name.as_str() == "0"
+    if let PatKind::Binding(BindingAnnotation::Unannotated, _, name1, None) = fields1[0].kind;
+    if name1.as_str() == "val";
     if let ExprKind::Break(ref destination, None) = arms1[1].body.kind;
+    if destination.label.is_none();
     if let PatKind::Struct(ref path7, ref fields2, false) = arms1[1].pat.kind;
     if matches!(path7, QPath::LangItem(LangItem::OptionNone, _));
     if fields2.len() == 0;
-    // unimplemented: field checks
     if let StmtKind::Local(ref local1) = body.stmts[2].kind;
     if let Some(ref init) = local1.init;
     if let ExprKind::Path(ref path8) = init.kind;
     if match_qpath(path8, &["__next"]);
-    if let PatKind::Binding(BindingAnnotation::Unannotated, _, name1, None) = local1.pat.kind;
-    if name1.as_str() == "y";
+    if let PatKind::Binding(BindingAnnotation::Unannotated, _, name2, None) = local1.pat.kind;
+    if name2.as_str() == "y";
     if let StmtKind::Expr(ref e1, _) = body.stmts[3].kind
     if let ExprKind::Block(ref block) = e1.kind;
     if block.stmts.len() == 1;
     if let StmtKind::Local(ref local2) = block.stmts[0].kind;
     if let Some(ref init1) = local2.init;
     if let ExprKind::Path(ref path9) = init1.kind;
     if match_qpath(path9, &["y"]);
-    if let PatKind::Binding(BindingAnnotation::Unannotated, _, name2, None) = local2.pat.kind;
-    if name2.as_str() == "z";
+    if let PatKind::Binding(BindingAnnotation::Unannotated, _, name3, None) = local2.pat.kind;
+    if name3.as_str() == "z";
     if block.expr.is_none();
     if body.expr.is_none();
-    if let PatKind::Binding(BindingAnnotation::Mutable, _, name3, None) = arms[0].pat.kind;
-    if name3.as_str() == "iter";
+    if label.is_none();
+    if let PatKind::Binding(BindingAnnotation::Mutable, _, name4, None) = arms[0].pat.kind;
+    if name4.as_str() == "iter";
+    then {
+        // report your lint here
+    }
+}
+if_chain! {
+    if let ExprKind::DropTemps(ref expr) = expr.kind;
+    if let ExprKind::Match(ref expr1, ref arms, MatchSource::ForLoopDesugar) = expr.kind;
```

Comment:
While you're at it, can you integrate `higher`?
```suggestion
    if let Some(ForLoop { pat, arg, body, .. }) = higher::ForLoop::parse(expr);
```

Reply:
Is author supposed to print the pattern as it would be in `rustc_ast` or `rust_hir`? I made `ExprKind::Closure` print as it would be in `rustc_ast`, but `ExprKind::Match` is printed as it is in `rustc_hir`.

Reply:
Author outputs code for matching the HIR.